File: androidApp/src/main/java/com/well/androidApp/ui/composableScreens/call/drawing/_DrawingPanel.kt
package com.well.androidApp.ui.composableScreens.call.drawing

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.layout.onSizeChanged
import com.well.androidApp.ui.composableScreens.πExt.toColor
import com.well.serverModels.Size
import com.well.sharedMobile.puerh.call.drawing.DrawingFeature.Msg
import com.well.sharedMobile.puerh.call.drawing.DrawingFeature.State
import com.google.accompanist.coil.CoilImage

@Composable
fun DrawingPanel(
    state: State,
    modifier: Modifier,
    listener: (Msg) -> Unit,
) = Column(
    modifier = modifier
        .fillMaxSize()
) {
    TopPanel(
        state = state,
        listener = listener,
        modifier = Modifier
            .fillMaxWidth()
    )
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .weight(1F)
    ) {
        BoxWithConstraints(
            modifier = Modifier
                .fillMaxSize()
                .onSizeChanged {
                    listener(Msg.UpdateLocalImageContainerSize(Size(it.width, it.height)))
                })
        {
            state.image?.let {
                CoilImage(
                    it.coilDataAny,
                    contentDescription = null,
                    contentScale = ContentScale.Fit,
                    modifier = Modifier
                        .fillMaxSize()
                        .background(Color.Black)
                )
                DrawingContent(
                    state, listener,
                    modifier = Modifier.fillMaxSize()
                )
            }
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
        ) {
            Spacer(modifier = Modifier.weight(1f))
            val lineWidth = remember { mutableStateOf(state.lineWidth) }
            WidthSlider(
                lineWidth,
                previewThumbColor = state.currentColor.toColor(),
                thumbRadiusRange = State.lineWidthRange,
                onValueChangeEnd = {
                    listener(Msg.UpdateLineWidth(lineWidth.value))
                },
                modifier = Modifier
                    .weight(4f)
                    .fillMaxWidth(0.5F)
            )
            Spacer(modifier = Modifier.weight(1f))
        }
    }
    BottomPanel(
        state = state,
        listener = listener,
    )
} // Column

File: iosApp/Shared/Screens/Call/ImageSharing/DrawingPanel.swift
//
// Created by Philip Dukhov on 1/9/21.
// Copyright (c) 2021 Well. All rights reserved.
//

import SwiftUI
import SharedMobile

struct DrawingPanel: View {
    typealias Listener = (DrawingFeature.Msg) -> Void
    private let geometry: GeometryProxy
    private let state: DrawingFeature.State
    private let listener: Listener
    private let back: () -> Void

    init(
        geometry: GeometryProxy,
        state: DrawingFeature.State,
        listener: @escaping Listener,
        back: @escaping () -> Void
    ) {
        self.geometry = geometry
        self.state = state
        self.listener = listener
        self.back = back
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            topPanel()
                .background(Color.black)
            ZStack(alignment: Alignment(horizontal: .leading, vertical: .center)) {
                GeometryReader { geometry in
                    ZStack(alignment: .leading) {
                        if let image = state.image {
                            Image(uiImage: image.uiImage)
                                .resizable()
                                .aspectRatio(contentMode: .fit)
                                .fillMaxSize()
                                .background(Color.black)
                            DrawingContent(
                                state: state,
                                enabled: true,
                                listener: listener
                            )
                        }
                    }.fillMaxSize()
                    .onAppear {
                        listener(
                            DrawingFeature.MsgUpdateLocalImageContainerSize(
                                size: geometry.size.toSize()
                            )
                        )
                    }
                }
                WidthSlider(
                    value: state.lineWidth,
                    range: DrawingFeature.StateCompanion().lineWidthRange.toClosedRange(),
                    color: state.currentColor.toColor()
                ) {
                    listener(
                        DrawingFeature.MsgUpdateLineWidth(lineWidth: $0)
                    )
                }.frame(width: geometry.size.width / 2, height: geometry.size.height / 2)
            }
            bottomPanel()
        }
        .foregroundColor(.white)
    }

    @ViewBuilder
    func topPanel() -> some View {
        ZStack(alignment: .center) {
            HStack {
                Control(Image.systemChevronLeft, action: back)
                Spacer()
                Control(Image(systemName: "trash.circle").font(.system(size: 20)), enabled: state.clearAvailable) {
                    listener(DrawingFeature.MsgLocalClear(saveHistory: true))
                }
                Control(Image(systemName: "photo").font(.system(size: 20))) {
                    listener(DrawingFeature.MsgRequestImageUpdate())
                }
            }
            HStack {
                Control(Image(systemName: "arrow.uturn.left"), enabled: state.undoAvailable) {
                    listener(DrawingFeature.MsgUndo())
                }
                Control(Image(systemName: "arrow.uturn.right"), enabled: state.redoAvailable) {
                    listener(DrawingFeature.MsgRedo())
                }
            }
        }.padding(.top, geometry.safeAreaInsets.top)
    }

    @ViewBuilder
    func bottomPanel() -> some View {
        let drawingColors = DrawingFeature.StateCompanion().drawingColors
        let spacing: CGFloat = 5
        let padding: CGFloat = 8
        Group {
            HStack(spacing: spacing) {
                ForEach(drawingColors, id: \.self) { color in
                    let selected = state.currentColor == color
                    Circle()
                        .stroke(lineWidth: 2)
                        .foregroundColor(selected ? .blue : .clear)
                        .scaledToFit()
                        .overlay(
                            Circle()
                                .foregroundColor(color.toColor())
                                .padding(3)
                        )
                        .onTapGesture {
                            listener(DrawingFeature.MsgUpdateColor(color: color))
                        }
                        .allowsHitTesting(!selected)
                }
            }
                .padding([.trailing, .leading, .top], padding)
                .padding(.bottom, padding + geometry.safeAreaInsets.bottom)
        }
        .backgroundColorKMM(.companion.MineShaft)
    }
}
